src/util/PendlePTHelper.sol

```diff
@@ -0,0 +1,281 @@
+        uint256 ytBalBefore = yt.balanceOf(address(this));
+        (bool success, ) = router.call(callData);
+        if (!success) revert PendleSwapFailed();
+
+        uint256 ptBal = pt.balanceOf(address(this)) - ptBalBefore;
+
+        if (ptBal < minMint) revert InsufficientPT();
+        if (recipient != address(this)) pt.safeTransfer(recipient, ptBal);
+        // Send YT to user if specified
+        if (ytRecipient != address(0)) {
+            uint256 ytBalMinted = yt.balanceOf(address(this)) - ytBalBefore;
+            if (ytBalMinted < minMint) revert InsufficientYT();
+            yt.safeTransfer(ytRecipient, ytBalMinted);
+        }
+
+        return ptBal;
+    }
+
+    /**
+     * @notice Redeems PT token for DOLA.
+     * @dev Used by the ALE but can be called by anyone. Carefully review input data for Pendle API.
+     * The receiver in Pendle API has to be set same as the recipient. If a REDEEM is performed, include a ytProvider with enough allowance.
+     * @param amount The amount of PT token to be redeemed (and YT if specified).
+     * @param data Encoded address of the market, minimum amount of DOLA to receive, ytProvider, and Pendle callData.
+     * @return dolaAmount The amount of DOLA redeemed.
+     */
+    function transformFromCollateral(
+        uint256 amount,
+        bytes calldata data
+    ) external override returns (uint256 dolaAmount) {
+        dolaAmount = transformFromCollateral(amount, msg.sender, data);
+    }
+
+    /**
+     * @notice Redeems Collateral for DOLA.
+     * @dev The receiver in Pendle API has to be set same as the recipient. If a REDEEM is performed, include a ytProvider with enough allowance.
+     * @param amount The amount of PT Token to be redeemed (and YT if specified).
+     * @param recipient The address to which the underlying token is transferred.
+     * @param data Encoded address of the market, minimum amount of DOLA to receive for the recipient, ytProvider, and Pendle callData.
+     * @return dolaAmount The amount of DOLA redeemed.
+     */
+    function transformFromCollateral(
+        uint256 amount,
+        address recipient,
+        bytes calldata data
+    ) public override returns (uint256 dolaAmount) {
+        (
+            address market,
+            uint256 minOut, // Minimum amount of DOLA to receive
+            address ytProvider,
+            bytes memory callData
+        ) = abi.decode(data, (address, uint256, address, bytes));
+        _revertIfMarketNotSet(market);
+
+        if (ytProvider != address(0)) {
+            IERC20 yt = IERC20(markets[market].yt);
+            yt.safeTransferFrom(ytProvider, address(this), amount);
+            yt.approve(router, amount);
+        }
+
+        IERC20 pt = IERC20(markets[market].pt);
+        pt.safeTransferFrom(msg.sender, address(this), amount);
+        pt.approve(router, amount);
+
+        uint256 dolaBal = DOLA.balanceOf(recipient);
+        (bool success, ) = router.call(callData);
```

Comment:
I'm assuming that in cases where the ytProvider is set, the router will know how to just redeem the PT & YT for collateral, while in cases where only PT tokens are provided, it will be able to either:
1. Sell the PT for collateral if before redemption date.
2. Redeem the PT for collateral if after the redemption date.